File: analysis/concentracao_fornecedores.py
from typing import Any

import pandas as pd
from sqlalchemy import bindparam, text

from analysis.constants import FORNECEDORES_NATUREZA_MAP


def run(conn: Any, year: int, empresa_id: str | None = None) -> dict:
    empresa_clause = "AND f.empresa = :empresa" if empresa_id else ""
    params: dict = {"ano": year}
    if empresa_id:
        params["empresa"] = empresa_id

    sql = text(
        f"""
        SELECT f.codigo, f.descricao, f.empenhado
        FROM despesas_por_fornecedor f
        LEFT JOIN despesas_gerais g
          ON f.ano = g.ano
          AND f.descricao = g.nomefor
        WHERE f.ano = :ano
        {empresa_clause}
        AND g.elemento IN :elementos
        """
    ).bindparams(
        bindparam("elementos", expanding=True, value=list(FORNECEDORES_NATUREZA_MAP.keys())),
    )

    df = pd.read_sql_query(sql, conn, params=params)
    df["empenhado"] = pd.to_numeric(df["empenhado"].str.replace(",", "."), errors="coerce").fillna(0)
    df = df.groupby(["codigo", "descricao"], as_index=False)["empenhado"].sum()
    total = df["empenhado"].sum()
    df["percentual"] = df["empenhado"] / total * 100 if total > 0 else 0
    top10 = df.nlargest(10, "empenhado").reset_index(drop=True)

    shares = df["empenhado"] / total if total > 0 else df["empenhado"] * 0
    hhi = float((shares**2).sum() * 10000)

    linha_dominante = df[df["percentual"] > 40]
    dominante = linha_dominante.iloc[0]["descricao"] if not linha_dominante.empty else None

    # total_all inclui entradas "E OUTROS" excluídas do top10 para que a fatia "Outros" do piechart seja precisa
    df_all = pd.read_sql_query(
        text(
            f"""
            SELECT f.empenhado
            FROM despesas_por_fornecedor f
            LEFT JOIN despesas_gerais g
              ON f.ano = g.ano
              AND f.descricao = g.nomefor
            WHERE f.ano = :ano
            {empresa_clause}
            AND g.elemento IN :elementos
            """
        ).bindparams(
            bindparam("elementos", expanding=True, value=list(FORNECEDORES_NATUREZA_MAP.keys())),
        ),
        conn,
        params=params,
    )
    df_all["empenhado"] = pd.to_numeric(df_all["empenhado"].str.replace(",", "."), errors="coerce").fillna(0)
    total_all = float(df_all["empenhado"].sum())

    return {"top10": top10, "hhi": hhi, "dominante": dominante, "total_all": total_all}


def piechart_concentracao(top10: pd.DataFrame, total_all: float) -> pd.DataFrame:
    outros = total_all - float(top10["empenhado"].sum())
    slices = top10[["descricao", "empenhado"]].rename(columns={"descricao": "Fornecedor"})
    if outros > 0:
        slices = pd.concat([slices, pd.DataFrame({"Fornecedor": ["Outros"], "empenhado": [outros]})])
    return slices


def hhi_por_ano(conn: Any, years: list[int], empresa_id: str | None = None) -> dict[int, float]:
    return {year: run(conn, year, empresa_id=empresa_id)["hhi"] for year in years}

File: packages/db/src/queries/concentracao_fornecedores.ts
import { sql } from "kysely";
import { db } from "../client";

export interface ItemFornecedorTop {
  codigo: string;
  descricao: string;
  empenhado: number;
  percentual: number;
}

export interface ConcentracaoResult {
  top10: ItemFornecedorTop[];
  hhi: number;
  dominante: string | null;
  total_all: number;
}

const ELEMENTOS_VAL = ["30", "33", "35", "36", "39", "40", "51", "52"];

export async function getConcentracaoFornecedores(
  year: number,
  empresaIds?: string[] | null,
): Promise<ConcentracaoResult> {
  try {
    let q = sql`
      SELECT f.codigo, f.descricao, f.empenhado
      FROM fct_despesas_por_fornecedor f
      LEFT JOIN fct_despesas g
        ON f.ano = g.ano
        AND f.descricao = g.fornecedor_nome
      WHERE f.ano = ${year}
        AND g.elemento = ANY(${ELEMENTOS_VAL})
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND f.empresa = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);
    const rows = (res.rows as any[]) || [];

    const map: Record<
      string,
      { codigo: string; descricao: string; empenhado: number }
    > = {};
    let total = 0;

    for (const r of rows) {
      const cod = String(r.codigo ?? "");
      const desc = String(r.descricao ?? "");
      const key = `${cod}__${desc}`;
      const emp = parseFloat(String(r.empenhado ?? "0").replace(",", ".")) || 0;

      if (!map[key]) {
        map[key] = { codigo: cod, descricao: desc, empenhado: 0 };
      }
      map[key].empenhado += emp;
      total += emp;
    }

    const items = Object.values(map).map((i) => ({
      ...i,
      percentual: total > 0 ? (i.empenhado / total) * 100 : 0,
    }));

    items.sort((a, b) => b.empenhado - a.empenhado);
    const top10 = items.slice(0, 10);

    const sumHHI = items.reduce((acc, i) => {
      const share = total > 0 ? i.empenhado / total : 0;
      return acc + share * share;
    }, 0);
    const hhi = sumHHI * 10000;

    const domItem = items.find((i) => i.percentual > 40);
    const dominante = domItem ? domItem.descricao : null;

    return { top10, hhi, dominante, total_all: total };
  } catch {
    return { top10: [], hhi: 0, dominante: null, total_all: 0 };
  }
}
